include/srf/pubsub/publisher.hpp

```diff
@@ -0,0 +1,148 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "srf/channel/ingress.hpp"
+#include "srf/codable/encoded_object.hpp"
+#include "srf/control_plane/subscription_service_forwarder.hpp"
+#include "srf/node/edge_builder.hpp"
+#include "srf/node/operators/unique_operator.hpp"
+#include "srf/node/source_channel.hpp"
+#include "srf/pubsub/api.hpp"
+#include "srf/runtime/api.hpp"
+#include "srf/runtime/remote_descriptor.hpp"
+#include "srf/utils/macros.hpp"
+
+namespace srf::pubsub {
+
+/**
+ * @brief Publishes an object T which will be received by one more more Subscribers.
+ *
+ * This object is both directly writeable, but also connectable to multiple upstream sources of T. By its nature as an
+ * operator, forward progress is performed not by a progress engine, but rather by the callers of await_write or by the
+ * execution context driving forward progress along the edge.
+ *
+ * Publisher<T> Data Path:
+ * [T] -> EncodedObject<T> -> EncodedStorage -> RemoteDescriptor -> Transient Buffer -> Data Plane Tagged Send
+ *
+ * Subscriber<T> Data Path:
+ * Data Plane Tagged Received -> Transient Buffer -> RemoteDescriptor -> Subscriber/Source<T> ->
+ *
+ * Subscriber<RemoteDescriptor> Data Path:
+ * Data Plane Tagged Received -> Transient Buffer -> RemoteDescriptor -> Subscriber/Source<RemoteDescriptor> ->
+ */
+template <typename T>
+class Publisher final : public control_plane::SubscriptionServiceForwarder,
+                        public node::UniqueOperator<T>,
+                        public channel::Ingress<T>
+{
+  public:
+    static std::unique_ptr<Publisher> create(std::string name,
+                                             const PublisherPolicy& policy,
+                                             runtime::IPartition& partition)
+    {
+        return std::unique_ptr<Publisher>{new Publisher(partition.make_publisher_service(name, policy))};
+    }
+
+    ~Publisher() final
+    {
+        request_stop();
+        await_join();
+    }
+
+    DELETE_COPYABILITY(Publisher);
+    DELETE_MOVEABILITY(Publisher);
+
+    // [Ingress<T>] publish T by capturing it as an encoded object, then pushing that encoded object to the internal
+    // publisher
+    channel::Status await_write(T&& data) final;
+
+    void await_start() final
+    {
+        // form a persistent connection to the operator
+        // data flowing in from operator edges are forwarded to the public await_write
+        m_persistent_channel = std::make_unique<srf::node::SourceChannelWriteable<T>>();
+        srf::node::make_edge(*m_persistent_channel, *this);
```

Comment:
I dont love this. Why is the `m_persistent_channel` needed here?

Reply:
Publisher is both directly writable and upstream connectable. 

Because of that, we can not just immediate request a control plane drop service on this tag. 

We have to wait to make the control plane drop request until after all upstream have been disconnected AND a call to request stop have been made on the object. 

We use are using the persistent channel model to do this. Initialize the persistent channel.  Upstream ingress acceptors can connect. Request stop drops the persistent connection. When the last connection is released the control plane messsge is issued.

Reply:
After discussion, I'm fine with this for now.